Deploy kube-dns with cluster-proportional-autoscaler

MrHohn opened the pull request:
This PR integrates [cluster-proportional-autoscaler](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler) with kube-dns for DNS horizontal autoscaling. 

Fixes #28648 and #27781.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33239)

<!-- Reviewable:end -->

MrHohn commented:
@k8s-bot gce e2e test this issue:#31561

MrHohn commented:
@k8s-bot gke smoke e2e test this issue:#28721

MrHohn commented:
@k8s-bot kubemark test this issue: #32789

MrHohn commented:
@k8s-bot gke e2e test this issue:#28721

MrHohn commented:
@k8s-bot test this please

roberthbailey commented:
Review status: 0 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[cluster/addons/dns/Makefile, line 32 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KSNMZVCKSytaNnVvZUk:-KSNMZVCKSytaNnVvZUl:bo01n07) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/Makefile#L32)):*
> ```Makefile
> 	sed -f transforms2sed.sed $<  | sed s/__SOURCE_FILENAME__/$</g > $@
> 
> transform: skydns-rc.yaml.in skydns-rc.yaml.sed skydns-svc.yaml.in skydns-svc.yaml.sed kubedns-autoscaler-configmap.yaml.in kubedns-autoscaler-configmap.yaml.sed kubedns-autoscaler-deployment.yaml.in kubedns-autoscaler-deployment.yaml.sed
> ```

I don't see any parameterized variables in any of the new files that warrant the base -> yaml.in and base -> yaml.sed transformations. Until we need variable substitution, it might be simpler to just have kubedns-autoscaler-configmap.yaml and kubedns-autoscaler-deployment.yaml as plain yaml files. 

---

*[cluster/images/hyperkube/static-pods/addon-manager-multinode.json, line 14 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KSNN1Tm1lneqdA4cZmq:-KSNN1Tm1lneqdA4cZmr:b-n9knz6) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/images/hyperkube/static-pods/addon-manager-multinode.json#L14)):*
> ```JSON
>       {
>         "name": "kube-addon-manager",
>         "image": "gcr.io/zihongz-kubernetes-codelab/kube-addon-manager-ARCH:v5.2",
> ```

Did you intend to commit this change?

---

*[cluster/images/hyperkube/static-pods/addon-manager-singlenode.json, line 14 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KSNN7OWmZ9Pj6NFKME4:-KSNN7OWmZ9Pj6NFKME5:b1p0oul) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/images/hyperkube/static-pods/addon-manager-singlenode.json#L14)):*
> ```JSON
>       {
>         "name": "kube-addon-manager",
>         "image": "gcr.io/zihongz-kubernetes-codelab/kube-addon-manager-ARCH:v5.2",
> ```

And this one?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239)*
<!-- Sent from Reviewable.io -->

MrHohn commented:
Review status: 0 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[cluster/addons/dns/Makefile, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KSNMZVCKSytaNnVvZUk:-KSODIIpkoMcWSNmeeLN:bud4tbg) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/Makefile#L32)):*
<details><summary><i>Previously, roberthbailey (Robert Bailey) wrote…</i></summary>
> I don't see any parameterized variables in any of the new files that warrant the base -> yaml.in and base -> yaml.sed transformations. Until we need variable substitution, it might be simpler to just have kubedns-autoscaler-configmap.yaml and kubedns-autoscaler-deployment.yaml as plain yaml files. 
</details>
You are right, I will change it.

---

*[cluster/images/hyperkube/static-pods/addon-manager-multinode.json, line 14 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KSNN1Tm1lneqdA4cZmq:-KSODR1OD-8WO-TFKyyG:bcv8xrz) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/images/hyperkube/static-pods/addon-manager-multinode.json#L14)):*
<details><summary><i>Previously, roberthbailey (Robert Bailey) wrote…</i></summary>
> Did you intend to commit this change?
</details>
No. I'm not sure when would the new version Addon Manager be pushed (to support ConfigMap). This private one is for testing only.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239)*
<!-- Sent from Reviewable.io -->

MrHohn commented:
@bowei

MrHohn commented:
For now the ConfigMap params (`cores_to_replicas_map` and `nodes_to_replicas_map`) in used for kube-dns are not well examined and tested.

As @bowei indicates, we should regenerate these params based on some real metrics.

thockin commented:
Review status: 0 of 20 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cluster/addons/dns/kubedns-autoscaler-configmap.yaml.base, line 32 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD2sjGaRjktctIZaEa:-KTD2sjHCIbu0HT-mmAi:b-c3dtjq) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-configmap.yaml.base#L32)):*
> ```
>       [
>         [ 1,1 ],
>         [ 4,2 ],
> ```

We don't need a 2nd replica for 4 cores.  Maybe for 64 cores

---

*[cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base, line 20 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD3LHSliBNUbEYzPDd:-KTD3LHTkvFqAVXTU-Ev:b-cwg3bp) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base#L20)):*
> ```
> kind: Deployment
> metadata:
>   name: kube-dns-cluster-proportional-autoscaler-v1
> ```

You don't need the "v1" gunk in the name for a deployment.

---

*[cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base, line 24 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD3RbUUlme7e6clO0V:-KTD3RbUUlme7e6clO0W:bpblchk) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base#L24)):*
> ```
>   labels:
>     k8s-app: kubedns-autoscaler
>     version: v1
> ```

I don't think you need this either

---

*[cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base, line 38 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD3hSHuBv-8MSSOFci:-KTD3hSHuBv-8MSSOFcj:b-e1hi86) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base#L38)):*
> ```
>         image: gcr.io/zihongz-kubernetes-codelab/cluster-proportional-autoscaler:0.6
>         resources:
>           limits:
> ```

I would leave limits off, and let them default.  I especially would not do burstable for memory.  Get a sense of how much you think this really needs to set request properly.

---

*[cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD46OdHHnnL734unXW:-KTD46OdHHnnL734unXX:b-e3eyqc) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-deployment.yaml.base#L48)):*
> ```
>           - --namespace=kube-system
>           - --configmap=kube-dns-cluster-proportional-autoscaler-params
>           - --target=rc/kube-dns-v19
> ```

This means we have to update this file when we update DNS.  The only maybe-better answer I see is to set a selector here, and I don't think we want that.  Open to ideas.

---

*[cluster/addons/dns/kubedns-autoscaler-deployment.yaml.sed, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD4RT0o2gW_IJoBGjM:-KTD4RT1MRn2rvNro6e7:bigp551) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/addons/dns/kubedns-autoscaler-deployment.yaml.sed#L1)):*
> ```
> # Copyright 2016 The Kubernetes Authors.
> ```

I don't think you need the in/base/sed for these files - they don't have any templated content.  Just yaml files should suffice.

---

*[cluster/saltbase/salt/kube-addons/init.sls, line 95 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD5FpCu3QdQvPTsWqX:-KTD5FpCu3QdQvPTsWqY:bqg70o4) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/saltbase/salt/kube-addons/init.sls#L95)):*
> ```Scheme
>   file.managed:
>     - source: salt://kube-addons/dns/kubedns-autoscaler-configmap.yaml.in
>     - template: jinja
> ```

This is not a jinja file, so it doesn't need this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239)*
<!-- Sent from Reviewable.io -->

thockin commented:
Is the hscaler itself committed?

---

Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-:-KTD5cNSJrH0Bb5Q0kvb:b-sdrrt0)*
<!-- Sent from Reviewable.io -->

MrHohn commented:
Thanks for all the comments. Not yet, actually that [PR](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/pull/3) is the one waiting for review.

MrHohn commented:
The new commits deploy kube-dns with cluster-proportional-autoscaler by making the autoscaler itself as an addon. This feature is turned on by default for gci and container-VM on GCE.

Notice that the this autoscaler still utilizes ConfigMap to retrieve autoscaling parameters, which is passed into the autoscaler pod by startup. This ConfigMap itself is not an addon, and Addon Manager will not reconcile it. Hence customers would be able to modify the autoscaling parameters. If this ConfigMap got deleted accidentally, autoscaler will manage to re-create it according to the default parameters passed in.

Beside, a basic e2e test for this autoscaling feature is added. It covers cases like cluster size changed, parameters changed, ConfigMap got deleted, autoscaler pod got deleted, etc.

@thockin @bowei

MrHohn commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,41 @@
+      containers:
+      - name: autoscaler
+        image: gcr.io/google_containers/cluster-proportional-autoscaler-amd64:0.8.5
+        imagePullPolicy: Always
```
Sorry, will remove this line.

MrHohn commented:
@rmmh Sorry to bother. Ran `./hack/update_owners.py`, but `./hack/verify-test-owners.sh` still failed on Jenkins.

MrHohn commented:
@k8s-bot node e2e test this, issue: #34990

MrHohn commented:
@k8s-bot node e2e test this, issue: #34206

MrHohn commented:
@k8s-bot verify test this, issue:#IGNORE

roberthbailey commented on cluster/gce/config-default.sh:
```diff
@@ -167,3 +167,6 @@ EVICTION_HARD="${EVICTION_HARD:-memory.available<100Mi,nodefs.available<10%,node
 SCHEDULING_ALGORITHM_PROVIDER="${SCHEDULING_ALGORITHM_PROVIDER:-}"
+
+# Optional: Enable DNS horizontal autoscaler
+ENABLE_DNS_HORIZONTAL_AUTOSCALER="${KUBE_ENABLE_DNS_HORIZONTAL_AUTOSCALER:-true}"
```
Why is this true when the default in cluster/common.sh is false? Is the intent just to enable it by default on gce?

@fabioy

MrHohn replied:
I was intended to only make it a default on gce, but not sure whether I should do this. @thockin

roberthbailey commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,40 @@
+          - --namespace=kube-system
+          - --configmap=kube-dns-autoscaler-params
+          - --mode=linear
+          - --target=ReplicationController/kube-dns-v20
```
The target will need to be changed whenever the name of the dns rc changes. can you add a comment to the DNS RC that points people to this file so that we remember to update them at the same time?

MrHohn replied:
Sure, will do that.

thockin replied:
If the changeover to deployments is finished, this gets better

thockin replied:
an alternative would be a selector here.

MrHohn commented:
@rmmh Sure, thanks for fixing!

thockin commented:
re: default - I want it on by default. but only if you're really confident in that (and have good reasons to be confident :)

thockin commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,40 @@
+  template:
+    metadata:
+      labels:
+        k8s-app: kube-dns-autoscaler
```
DNS is marked "critical" - should this be, also?  What happens if the cluster is full and this is not running?  I guess as the cluster scales up, it will have opportunities to run.  I'd still consider it critical

thockin commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: kube-dns-autoscaler
```
Curious: why make a new dir for this, rather than dropping it in with DNS files?

roberthbailey commented:
> re: default - I want it on by default. but only if you're really confident in that (and have good reasons to be confident :)

On *everywhere* by default or just on GCE by default? I think if we are confident we should default it on everywhere (and add a release note).

MrHohn commented:
Review status: 0 of 10 files reviewed at latest revision, 15 unresolved discussions.

---

*[cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml, line 32 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KV6aud_bdccvmHvnoFR-r3-32:-KV_xQM9JWjmB8_6nK3l:b-4gm03z) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0a4e03d83182c62c92ee221e9010a38bd936874/cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml#L32)):*
<details><summary><i>Previously, MrHohn (Zihong Zheng) wrote…</i></summary>
> Sorry, will remove this line.
</details>
`imagePullPolicy: Always` was removed.

---

*[cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml, line 18 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KV6aud_bdccvmHvnoFR-r5-18:-KV_wW7L8-ltHeIAFhCS:b-nm9v6b) ([raw file](https://github.com/kubernetes/kubernetes/blob/cdc593795994a55dbbe6cd561294ab2165fdd51c/cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml#L18)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> Curious: why make a new dir for this, rather than dropping it in with DNS files?
</details>
This make it easier to separate ENABLE_CLUSTER_DNS and ENABLE_DNS_HORIZONTAL_AUTOSCALER without re-construct the cloud configure scripts.

---

*[cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml, line 27 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KV6aud_bdccvmHvnoFR-r5-27:-KV_xCxbHPr6BibV7PMg:bufxrhv) ([raw file](https://github.com/kubernetes/kubernetes/blob/cdc593795994a55dbbe6cd561294ab2165fdd51c/cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml#L27)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> DNS is marked "critical" - should this be, also?  What happens if the cluster is full and this is not running?  I guess as the cluster scales up, it will have opportunities to run.  I'd still consider it critical
</details>
Make sense. Will add "critical".

---

*[cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml, line 37 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KV6aud_bdccvmHvnoFR-r5-37:-KV_xb3Fw9BynlgHGt_O:b-o85uvz) ([raw file](https://github.com/kubernetes/kubernetes/blob/cdc593795994a55dbbe6cd561294ab2165fdd51c/cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml#L37)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> an alternative would be a selector here.
</details>
Yeah, I remember you mentioned selector before but said we don't want that neither. Why?

---

*[cluster/saltbase/salt/kube-addons/init.sls, line 95 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239#-KTD5FpCu3QdQvPTsWqX:-KV_y9KCwHcFa7Aowh-r:bt2vr1b) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b819f1ab4679bc42465fde26da8bcaedb746b2a/cluster/saltbase/salt/kube-addons/init.sls#L95)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> This is not a jinja file, so it doesn't need this.
</details>
Fixed.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33239)*
<!-- Sent from Reviewable.io -->

MrHohn commented:
The new commits separates the new e2e tests into a fast part(could be run parallelly) and a slow part(put in [serial]) instead of putting them all in [serial]. The fast part of the e2e tests cost around 50 seconds to run, but in this case it would be run more frequently on CIs and make it easier for bug catching.

About `default`, for now I don't have a good reason to say we are really confident in it because even the e2e tests in this PR are not in yet. Besides, we have not finished large scale test nor dns performance measurement for this.

bprashanth commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,46 @@
+          - --configmap=kube-dns-autoscaler
+          - --mode=linear
+          # Should keep target in sync with cluster/addons/dns/skydns-rc.yaml.base
+          - --target=ReplicationController/kube-dns-v20
```
don't forget to change this after your other pr

bprashanth commented on cluster/addons/dns-horizontal-autoscaler/dns-horizontal-autoscaler.yaml:
```diff
@@ -0,0 +1,46 @@
+        scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
+    spec:
+      containers:
+      - name: autoscaler
```
if you don't set cpu/memory request you will get the crappy defaults, is this ok? suggest setting approximates from at least running this and using some tool like top (the more scientific the better of course)

bprashanth commented on cluster/gce/config-default.sh:
```diff
@@ -115,6 +115,9 @@ DNS_SERVER_IP="${KUBE_DNS_SERVER_IP:-10.0.0.10}"
 DNS_REPLICAS=1
 
+# Optional: Enable DNS horizontal autoscaler
+ENABLE_DNS_HORIZONTAL_AUTOSCALER="${KUBE_ENABLE_DNS_HORIZONTAL_AUTOSCALER:-true}"
```
you also need to update the gke configs: https://github.com/kubernetes/kubernetes/blob/master/cluster/gke/config-common.sh#L39 (follow up pr is fine if you prefer but I guess you need to specify limits in this one anyway). 
Do all other cloudproviders handle this gracefully? or do you have to pipe some env var down into them as well?

MrHohn committed:
Deployed DNS horizontal autoscaler as an addon

DNS horizontal autoscaling feature is turned on by default on gce.
The corresponding env var is piped into almost all other cloud
providers.

MrHohn commented:
The new commit sets cpu(20m)/memory(10Mi) request for the autoscaler pod. Limit is leaved unset.

GKE config is updated. Also piped down the corresponding env variable to other cloudproviders. But dns autoscaling is turned on by default only on gce.

MrHohn committed:
Adds e2e tests for DNS horizontal autoscaling feature

The e2e tests cover cases like cluster size changed, parameters
changed, ConfigMap got deleted, autoscaler pod got deleted, etc.
They are separated into a fast part(could be run parallelly) and
a slow part(put in [serial]). The fast part of the e2e tests cost
around 50 seconds to run.

MrHohn commented:
Tests passed this time.

bprashanth commented:
LGTM, make sure you do https://github.com/kubernetes/kubernetes/pull/33239/files#r86699869

bprashanth commented:
I'm bumping to P2 because we need this to go in before https://github.com/kubernetes/kubernetes/pull/36008, but I want to lgtm that one as it's the last day before code freeze

MrHohn commented:
Fixes #27781.
